proxy/src/scram/exchange.rs

```diff
@@ -75,46 +77,94 @@ impl<'a> Exchange<'a> {
     }
 }
 
-// copied from <https://github.com/neondatabase/rust-postgres/blob/20031d7a9ee1addeae6e0968e3899ae6bf01cee2/postgres-protocol/src/authentication/sasl.rs#L236-L248>
 async fn derive_client_key(
     pool: &ThreadPool,
     endpoint: EndpointIdInt,
     password: &[u8],
     salt: &[u8],
     iterations: u32,
 ) -> ScramKey {
-    let salted_password = pool
-        .spawn_job(endpoint, Pbkdf2::start(password, salt, iterations))
-        .await;
-
-    let make_key = |name| {
-        let key = Hmac::<Sha256>::new_from_slice(&salted_password)
-            .expect("HMAC is able to accept all key sizes")
-            .chain_update(name)
-            .finalize();
-
-        <[u8; 32]>::from(key.into_bytes())
-    };
-
-    make_key(b"Client Key").into()
+    pool.spawn_job(endpoint, Pbkdf2::start(password, salt, iterations))
+        .await
 }
 
 pub(crate) async fn exchange(
     pool: &ThreadPool,
     endpoint: EndpointIdInt,
+    role: RoleNameInt,
     secret: &ServerSecret,
     password: &[u8],
 ) -> sasl::Result<sasl::Outcome<super::ScramKey>> {
+    // hot path: let's check the threadpool cache
+    if let Some(cached) = pool.cache.get(&(endpoint, role)) {
+        // cached key is no longer valid.
+        if secret.is_password_invalid(&cached.client_key).into() {
+            debug!("invalidating cached password");
+            cached.invalidate();
+        } else if let Ok(client_key) = cached.verify(password) {
+            trace!("password validated from cache");
+            return Ok(sasl::Outcome::Success(client_key));
+        }
+    }
+
+    // slow path: full password verify.
     let salt = base64::decode(&secret.salt_base64)?;
     let client_key = derive_client_key(pool, endpoint, password, &salt, secret.iterations).await;
 
     if secret.is_password_invalid(&client_key).into() {
         Ok(sasl::Outcome::Failure("password doesn't match"))
     } else {
+        trace!("storing cached password");
+        pool.cache.insert_unit(
+            (endpoint, role),
+            ClientSecretEntry::new(password, client_key.clone()),
+        );
+
         Ok(sasl::Outcome::Success(client_key))
     }
 }
 
+#[derive(Clone)]
+pub struct ClientSecretEntry {
+    salt: [u8; 64],
+    hash: [u8; 64],
+    client_key: ScramKey,
+}
+
+impl Drop for ClientSecretEntry {
+    fn drop(&mut self) {
+        self.salt.zeroize();
+        self.hash.zeroize();
+    }
+}
+
+impl ClientSecretEntry {
+    fn new(password: &[u8], client_key: ScramKey) -> Self {
+        let mut salt = [0; 64];
+        rand::thread_rng().fill_bytes(&mut salt);
+
+        let mut hmac = hmac::Hmac::<sha2::Sha512>::new_from_slice(password)
+            .expect("HMAC is able to accept all key sizes");
```

Comment:
How about we use the salt as the key and hash the password instead? That way the key size is always known. I couldn't find it in the `hmac` crate implementation, but HMAC is supposed to hash the key if it's longer than the block size. Using the salt as the key avoids this step.